CodegenProposal.md

```diff
@@ -0,0 +1,1474 @@
+
+Instead of protocols, fragments are generated as concrete `SelectionSet` structs that any `SelectionSet` that contains the fragment can then be converted to. 
+
+**Fragment Definition:**
+
+```graphql
+fragment AnimalDetails on Animal {
+  species
+}
+```
+
+**Query Input:**
+
+```graphql
+query AllAnimalSpecies {
+  allAnimals {
+    ...AnimalDetails
+  }
+}
+```
+
+**Generated Output:**
+`AnimalDetails.Swift`
+
+```swift
+struct AnimalDetails: SelectionSet, Fragment {
+  static var __parentType: ParentType { .Interface(AnimalKingdomAPI.Animal.self) }
+  let data: ResponseDict
+
+  var species: String { data["species"] }
+}
+```
+
+`AllAnimalSpeciesQuery.swift` (`Animal` Object)
+
+```swift
+struct Animal: SelectionSet, HasFragments {
+    static var __parentType: ParentType { .Interface(AnimalKingdomAPI.Animal.self) }
+    let data: ResponseDict
+    
+    var species: String { data["species"] }
+      
+    struct Fragments: ResponseObject {
+      let data: ResponseDict
+
+      var animalDetails: AnimalDetails { _toFragment() }
+    }
+}
+```
+
+The query’s `Animal` struct conforms to `HasFragments`, which is a protocol that exposes a `fragments` property that exposes the nested `Fragments` struct. The fragments a `SelectionSet` contains are exposed in this `Fragments` struct via computed properties that utilize a `_toFragment()` helper function. This allows you to access the `AnimalDetails` fragment via `myAnimal.fragments.animalDetails`.
+
+### Merging Fragment Fields Into Parent `SelectionSet`
+
+In the above example you may note that the `species` field is accessible directly on the `Animal` object without having to access the `AnimalDetails` fragment first. This is because fields from fragments that have the same `__parentType` as the enclosing `SelectionSet` are automatically merged into the enclosing `SelectionSet`. 
+
+In previous versions of the Code Generation tool, this functionality was exposed via the mergeInFieldsFromFragmentSpreads option, which defaulted to `false`. Merging in fragment fields provides for easier consumption of generated models, but it also increases the size of the generated code. Because the size of the generated code is being dramatically reduced with the new Code Generation tooling, we have opted to always merge in fragment fields. If generated code size becomes a concern with the new Code Generation, adding an option to disable fragment merging may be considered in the future.
```

Comment:
The main issue with enabling `mergeInFieldsFromFragmentSpreads` is that codegen size grows exponentially with the depth of nested fragments. So even if the new codegen generates 10x less code, a deep query using a lot of fragments might still generate a huge amount of code.

Reply:
Yes, agreed. Because we generate dramatically less code, I feel that it's likely going to not be a problem. But if users complain about it, we can add that option in in the future. Codegen is such a massively scoped project, so we are trying to cut down to the MVP as much as possible, and then add more functionality from there.